### DISCUSSION TITLE
Make use of Jinja async_render in HA

### DISCUSSION THREAD

**Discussion by Petro31** [home-assistant/architecture#969] [General]
With the recent deprecation of the forecast attribute and the addition of returned values in service calls, we should visit the ability to call services directly from a Jinja template.

# Problem

Currently, we are limited to using automation variables to store information gathered from service calls.  This is great, however it is limited by variable scope inside an automation.  Meaning, you cannot call the service in an if-statement, repeat loop, choose, or any other nested action because the variable is not accessible outside.  This forces users to create multiple service calls with multiple variables just to piece the information together.  A great example of this is getting events from multiple calendars.  If you have 10 calendars, you'd need to make 10 service calls with 10 separate variables, all inside the root set of actions.

# Solution

Add async_render as an option for template rendering.  Add this next to the current implementation to keep backward compatibility while slowly rolling it out to integrations that need it.  Once complete, we can start adding new functionality that potentially allows users to call services directly inside a template.  This will allow users to make use of namespace, for loops, if statements to get multiple values from service calls.  This should also open up more doors for other functionality inside templates.

**Comment by balloob** [comment]
When we render a template, we currently do it inside a `callback` function. It means it's done within the same task that it got started in the event loop. This guarantees us that if we render 5 templates, that all of them are rendered against the same state of the system.

If we were to include async calls, a render of a template can now take any amount of time and our guarantee that all templates render against the same state no longer holds true. This becomes complicated when rendering a template entity with templates for name, icon and state. They can get out of sync! 

On top of that, we already have the script syntax. The script syntax allows users to call services and do a whole set of other async tasks. Allowing templates to do async things would turn it into a parallel script syntax. This will make things very complicated.

Lastly, Jinja is a template rendering engine. Not a programming language. Our script syntax already allows many programming constructs, we shouldn't let Jinja suffer the same faith. Because after all, if we want to program, shouldn't we just write Python?
